pkg/kvstore/etcd.go

```diff
@@ -0,0 +1,392 @@
+//
+// Copyright 2016 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package kvstore
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/common/types"
+
+	client "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	ctx "golang.org/x/net/context"
+)
+
+type EtcdClient struct {
+	cli         *client.Client
+	session     *concurrency.Session
+	sessionMU   sync.RWMutex
+	lockPaths   map[string]*sync.Mutex
+	lockPathsMU sync.Mutex
+}
+
+type EtcdLocker struct {
+	mutex     *concurrency.Mutex
+	localLock *sync.Mutex
+	path      string
+}
+
+func NewEtcdClient(config *client.Config, cfgPath string) (KVClient, error) {
+	var (
+		c   *client.Client
+		err error
+	)
+	if cfgPath != "" {
+		c, err = client.NewFromConfigFile(cfgPath)
+	} else if config != nil {
+		c, err = client.New(*config)
+	} else {
+		err = fmt.Errorf("empty configuration provided")
+	}
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Waiting for etcd client to be ready...")
+	s, err := concurrency.NewSession(c)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to contact etcd: %s", err)
+	}
+	log.Info("Etcd client ready")
+	ec := &EtcdClient{
+		cli:         c,
+		session:     s,
+		sessionMU:   sync.RWMutex{},
+		lockPaths:   map[string]*sync.Mutex{},
+		lockPathsMU: sync.Mutex{},
+	}
```

Comment:
Since you can now leverage the zero-values, this'd be equivalent:

```go
	ec := &EtcdClient{
		cli:         c,
		session:     s,
		lockPaths:   map[string]*sync.Mutex{},
	}
```

Which is admittedly more useful when the consumers of the API instantiate the struct themselves.